src/ast/passes/printer.h

```diff
@@ -1,72 +1,162 @@
 #pragma once
 
+#include <memory>
 #include <ostream>
 
 #include "ast/visitor.h"
 
 namespace bpftrace::ast {
 
-class Printer : public Visitor<Printer> {
+// BufferState is an internal type.
+struct BufferState;
+
+// Buffer is an internal type for output construction.
+class Buffer {
+public:
+  Buffer();
+  ~Buffer();
+  Buffer(Buffer &&) = default;
+  Buffer &operator=(Buffer &&) = default;
+  Buffer(const Buffer &) = delete;
+  Buffer &operator=(const Buffer &) = delete;
+
+  [[nodiscard]] Buffer &&append(Buffer &&other, size_t indent = 0);
+  [[nodiscard]] Buffer &&text(std::string str);
+  [[nodiscard]] Buffer &&metadata(
+      MetadataIndex metadata,
+      std::optional<size_t> min_vspace = std::nullopt);
+  [[nodiscard]] Buffer &&comment(std::string str);
+  [[nodiscard]] Buffer &&line_break();
+
+  size_t width() const;
+  size_t lines() const;
+
+private:
+  std::unique_ptr<BufferState> state;
+  friend std::ostream &operator<<(std::ostream &out, const Buffer &buffer);
+};
+
+// Buffer is directly printable, used below.
+std::ostream &operator<<(std::ostream &out, const Buffer &buffer);
+
+// Control how types should be formatted.
+enum FormatMode {
+  Minimal, // Only probes, no comments or spacing.
+  Full,    // Print with full comments and spacing.
```

Comment:
I saw a check for this but it doesn't seem like there is any way to invoke this mode. Is this in a follow up PR?